src/firestore/index.ts

```diff
@@ -96,13 +110,29 @@ export {
  *   service if no app is provided or the `Firestore` service associated with the
  *   provided app.
  */
-export function getFirestore(app?: App): Firestore {
-  if (typeof app === 'undefined') {
-    app = getApp();
-  }
-
+export function getFirestore(app: App): Firestore;
+/**
+ * @param databaseId
+ * @internal
+ */
+export function getFirestore(databaseId: string): Firestore;
+/**
+ * @param app
+ * @param databaseId
+ * @internal
+ */
+export function getFirestore(app: App, databaseId: string): Firestore;
+export function getFirestore(
+  appOrDatabaseId?: App | string,
+  optionalDatabaseId?: string
+): Firestore {
+  const app: App = typeof appOrDatabaseId === 'object' ? appOrDatabaseId : getApp();
```

Comment:
I think `type of null` is also `object`... we might want to check if `app !== null` as well. Alternatively, you can also move the implementations into the respective functions separately.

Reply:
They would need to bypass Typescript type checking for this to happen, since `appOrDatabaseId?: App | string`.

According to this doc, we should not guard against invalid types.
https://g3doc.corp.google.com/firebase/g3doc/aip/3270.md?cl=head#input-validation